presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -0,0 +1,282 @@
+import static com.facebook.airlift.http.client.Request.Builder.prepareGet;
+import static com.facebook.airlift.http.client.Request.Builder.preparePost;
+import static com.facebook.airlift.http.client.StaticBodyGenerator.createStaticBodyGenerator;
+import static com.facebook.presto.execution.TaskInfo.createInitialTask;
+import static com.facebook.presto.execution.buffer.OutputBuffers.BufferType.PARTITIONED;
+import static com.facebook.presto.execution.buffer.OutputBuffers.createInitialEmptyOutputBuffers;
+import static com.facebook.presto.server.RequestHelpers.setContentTypeHeaders;
+import static com.facebook.presto.server.smile.AdaptingJsonResponseHandler.createAdaptingJsonResponseHandler;
+import static com.facebook.presto.spi.StandardErrorCode.NATIVE_EXECUTION_TASK_ERROR;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * NativeExecutionTask provide the abstraction of executing tasks via c++ worker.
+ * The plan is to use it for native execution of Presto on Spark.
+ * The plan and splits is provided at creation of NativeExecutionTask, it doesn't
+ * support adding more splits during execution.
+ */
+public class NativeExecutionTask
+{
+    private static final Logger log = Logger.get(NativeExecutionTask.class);
+    private static final int GET_TASK_INFO_INTERVALS = 100;
+    private static final String LOCAL_URI = "http://127.0.0.1/v1/task/";
+    private final Session session;
+    private final TaskId taskId;
+    private final PlanFragment planFragment;
+    private final OutputBuffers outputBuffers;
+    private final HttpClient httpClient;
+    private final TableWriteInfo tableWriteInfo;
+    private final RemoteTaskStats stats;
+    private final List<TaskSource> sources;
+    private final Executor executor;
+    private final ScheduledExecutorService updateScheduledExecutor;
+    private final JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec;
+    private final JsonCodec<PlanFragment> planFragmentCodec;
+    private final JsonCodec<TaskInfo> taskInfoCodec;
+
+    @GuardedBy("this")
+    private TaskInfo taskInfo;
+    @GuardedBy("this")
+    private ScheduledFuture<?> scheduledFuture;
+
+    public NativeExecutionTask(
+            Session session,
+            TaskId taskId,
+            PlanFragment planFragment,
+            List<TaskSource> sources,
+            HttpClient httpClient,
+            TableWriteInfo tableWriteInfo,
+            RemoteTaskStats stats,
+            Executor executor,
+            ScheduledExecutorService updateScheduledExecutor,
+            JsonCodec<TaskUpdateRequest> taskUpdateRequestCodec,
+            JsonCodec<PlanFragment> planFragmentCodec,
+            JsonCodec<TaskInfo> taskInfoCodec)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(taskId, "taskId is null");
```

Comment:
can you do these when setting the variables instead o at the top.  It makes it easier to see what checks are happening for which variables.

e.g. 
this.taskId = requireNonNull(taskInfo, "taskId is null")